### PR TITLE
Changing behavior when a transform from odom->base_link fails

### PR DESCRIPTION
**Pull request by ayrton04**
@locusrobotics/robot-software-team

### CODE REVIEW

**Comment by locusroboticsdev** [comment]
[Fix karto transform issue](https://trello.com/c/2zVd6tE0/103-fix-karto-transform-issue)

File: src/slam_karto.cpp [hunk 1]
```diff
@@ -694,27 +695,41 @@ SlamKarto::addScan(karto::LaserRangeFinder* laser,
...
     // Compute the map->odom transform
+    tf::Transform cor_pose_trans(
+      tf::createQuaternionFromRPY(0, 0, corrected_pose.GetHeading()),
+      tf::Vector3(corrected_pose.GetX(), corrected_pose.GetY(), 0.0));
+
     tf::Stamped<tf::Pose> odom_to_map;
+
+    ros::Time transform_time = scan->header.stamp;
+
+    if (!tf_.canTransform(odom_frame_, base_frame_, scan->header.stamp))
```
Comment (tappan-at-git) [inline src/slam_karto.cpp]: Might be cleaner to use `transform_time` instead of repeating `scan->header.stamp` here.

File: src/slam_karto.cpp [hunk 2]
```diff
@@ -694,27 +695,41 @@ SlamKarto::addScan(karto::LaserRangeFinder* laser,
...
+    {
+      transform_time = ros::Time(0);
+    }
+
     try
     {
-      tf_.transformPose(odom_frame_,tf::Stamped<tf::Pose> (tf::Transform(tf::createQuaternionFromRPY(0, 0, corrected_pose.GetHeading()),
-                                                                    tf::Vector3(corrected_pose.GetX(), corrected_pose.GetY(), 0.0)).inverse(),
-                                                                    scan->header.stamp, base_frame_),odom_to_map);
+      tf::Transform cor_pose_trans(
```
Comment (tappan-at-git) [inline src/slam_karto.cpp]: You already defined this at 698.

File: src/slam_karto.cpp [hunk 3]
```diff
@@ -694,27 +695,41 @@ SlamKarto::addScan(karto::LaserRangeFinder* laser,
     karto::Pose2 corrected_pose = range_scan->GetCorrectedPose();
 
     // Compute the map->odom transform
+    tf::Transform cor_pose_trans(
+      tf::createQuaternionFromRPY(0, 0, corrected_pose.GetHeading()),
```
Comment (tappan-at-git) [inline src/slam_karto.cpp]: What frame is corrected_pose in? (I'm assuming it represents base->odom in some form, but I figured I'd check.)

Comment (ayrton04) [inline src/slam_karto.cpp]: Not entirely sure, but I didn't change that code. I just pulled it out of the conditional check below and put it in a variable instead. If the karto authors had this wrong, I think we'd see more trouble than we do.

File: src/slam_karto.cpp (hunk 1 above)
Comment (ayrton04) [inline src/slam_karto.cpp]: Agreed

File: src/slam_karto.cpp (hunk 2 above)
Comment (ayrton04) [inline src/slam_karto.cpp]: Yes, yes I did.

File: src/slam_karto.cpp (hunk 3 above)
Comment (tappan-at-git) [inline src/slam_karto.cpp]: Fair enough.

**Pull request merged**